### DISCUSSION TITLE
Why does ZChannel.toStream requires InErr, InElem and InDone to be 'Any'?

### DISCUSSION THREAD

**Discussion by pvgoran** [zio/zio#7201] [Q&A]
I have a channel that doesn't accept any input (neither errors nor elements nor the final value): `ZChannel[Env, Nothing, Nothing, Nothing, OutErr, OutElem, OutDone]`. I want to turn it into a `ZStream`:

```scala
val channel : ZChannel[Env, Nothing, Nothing, Nothing, OutErr, OutElem, OutDone] = ???
val stream : ZStream[Env, OutErr, OutElem] = channel.mapOut(Chunk(_)).toStream
```

I get an error "Cannot prove that Any <:< Nothing".

Why is it designed this way? A `ZStream` does not have any input by design, and my channel does not have any input due to its type, so turning the latter into the former should be a natural conversion.

What's worse, I don't see any type-safe way to transform my channel into a channel that can be converted to a stream. I can use `contramapInZIO` and `contramapZIO` to get a channel with `InElem`=`Any` and `InDone`=`Any`, but I `InErr` seems to be out of reach. So I may need to resort to `asInstanceOf` which is dangerous.

(Also see https://github.com/zio/zio/issues/7202.)

**Comment by vigoo** [comment]
You should represent the fact that the channel is not taking any input by using `Any` in the contravariant type parameters (`Env`, `InErr`, `InElem`, `InDone`).

**Reply by pvgoran** [reply]
Could you elaborate, please? Why is `Any` more appropriate than `Nothing` (which is the explicit "no input is possible" type, as far as I understand)?

**Reply by vigoo** [reply]
You can think of it as the input type is not bound (it can be `Any`thing) because it is not using it at all.

This also means that because of the contravariance, a channel of type `InElem=Any` is a subtype of any other channel with a specific input element type; you can use a channel that is not consuming any input in place of any other channel consuming input regardless of the input type, as it is never reading anything.

**Comment by adamgfraser** [comment]
A `ZChannel[Env, Nothing, Nothing, Nothing, OutErr, OutElem, OutDone]` is a channel that can never be run. It requires inputs that could never possibly exist. A `ZChannel[Env, Any, Any, Any, OutErr, OutElem, OutDone]` is a channel that can be run with any possible inputs, and thus does not do anything with those inputs because those inputs contain no useful information.

Another way to think about this is that `Nothing` is a subtype of every other type. So if I had a `ZChannel[Env, Nothing, Nothing, Nothing, OutErr, OutElem, OutDone]` that would be a subtype of a `ZChannel[Env, Foo, Foo, Foo, OutErr, OutElem, OutDone]`. If we could run a channel of type of `ZChannel[Env, Nothing, Nothing, Nothing, OutErr, OutElem, OutDone]` that would imply we could also run a channel of type `ZChannel[Env, Foo, Foo, Foo, OutErr, OutElem, OutDone]`. But that is not the case because the channel requires an input of type `Foo` and we have no way to produce that input.

You should represent your channel as a `ZChannel[Env, Any, Any, Any, OutErr, OutElem, OutDone]` and then it can naturally be converted to a stream by chunking its elements and ignoring its done value.
